### PR TITLE
Support hierarchical partition keys

### PR DESCRIPTION
**Pull request by thed24**
This method of extension should be backwards compatible with all existing usages of the trait whilst allowing users to opt into using the preview feature correctly.

### CODE REVIEW

**Comment by thed24** [comment]
@microsoft-github-policy-service agree [company="no company"]

**Comment by thed24** [comment]
@microsoft-github-policy-service agree

**Comment by thed24** [comment]
@microsoft-github-policy-service agree

File: sdk/data_cosmos/src/cosmos_entity.rs [hunk 1]
```diff
@@ -44,11 +44,20 @@ impl CosmosEntity for serde_json::Value {
...
+    // if pk is array, we dont need to wrap it in an array
+    // if it is not, we need to wrap it in an array
+
+    let serialized_string = serde_json::to_string(pk).context(
         azure_core::error::ErrorKind::DataConversion,
         "could not convert partition_key into String",
-    )
+    )?;
+
+    if serialized_string.starts_with('[') {
```
Comment (cataggar) [inline sdk/data_cosmos/src/cosmos_entity.rs:56]: Although this will work, can this be implemented without checking the serialization output?

**Review by cataggar — COMMENTED** [review]
Can a test be added?

File: sdk/data_cosmos/src/cosmos_entity.rs (hunk 1 above)
Comment (thed24) [reply sdk/data_cosmos/src/cosmos_entity.rs:56]: I tried changing some of the existing tests, although they relied on predefined `result` data that was asserted against so I left it be. Perhaps I can write some in-file tests?

The only way I could imagine achieving this without checking the serialized text is matching on the implementation of the generic type or something along those lines. This is method is more coupled, although safer as it works for all vector-like types or anything that serializes down to an array in JSON.

**Comment by cataggar** [comment]
I opened #1410, but closing this as it is stale.

**Pull request closed without merging**